plugins/net_plugin/net_plugin.cpp

```diff
@@ -2182,14 +2184,37 @@ namespace eosio {
               sync_next_expected_num < sync_last_requested_num );
    }
 
+   // called from c's connection strand
    bool sync_manager::is_sync_request_ahead_allowed(block_num_type blk_num) const REQUIRES(sync_mtx) {
       if (blk_num >= sync_last_requested_num) {
          // do not allow to get too far ahead (sync_fetch_span) of chain head
          // use chain head instead of fork head so we do not get too far ahead of applied blocks
-         uint32_t head = my_impl->get_chain_head_num();
-         if (blk_num < head + sync_fetch_span)
+         uint32_t head_num = my_impl->get_chain_head_num();
+         if (blk_num < head_num + sync_fetch_span)
             return true;
+
+         // might be in irreversible mode
+         controller& cc = my_impl->chain_plug->chain();
+         if (cc.get_read_mode() == db_read_mode::IRREVERSIBLE) {
+            // chain head == lib == fork_db_root in irreversible
+            auto forkdb_head = cc.fork_db_head();
+            if (forkdb_head.block_num() - head_num >= sync_fetch_span) {
```

Comment:
Add a small assert ensuring `forkdb_head.block_num() >= head_num`

Reply:
It is possible we catch forkdb in the middle of a fork switch. We could guard against that or let this if pass, worse case it allows an extra request of blocks. Although to avoid confusion likely better to make it `if (forkdb_head.block_num() > head_num && forkdb_head.block_num() - head_num >= sync_fetch_span) {`